File: CrossTimeDsp/Dsp/GainQ31.cs
using System;

namespace CrossTimeDsp.Dsp
{
    internal class GainQ31 : IFilter<int>
    {
        private Q31 scaleFactor;

        public GainQ31(double scaleFactor)
        {
            this.scaleFactor = new Q31(scaleFactor);
        }

        public void Filter(int[] block, int offset)
        {
            int maxSample = offset + Constant.FilterBlockSizeInInts;
            for (int sample = offset; sample < maxSample; ++sample)
            {
                block[sample] = (int)((this.scaleFactor * block[sample]) >> this.scaleFactor.FractionalBits);
            }
        }

        public void FilterReverse(int[] block, int offset)
        {
            this.Filter(block, offset);
        }
    }
}

File: Native/GainQ31.cpp
#include "stdafx.h"
#include "Constant.h"
#include "GainQ31.h"

namespace CrossTimeDsp::Dsp
{
	GainQ31::GainQ31(double scaleFactor)
	{
		this->scaleFactor = new Q31(scaleFactor);
	}

	GainQ31::~GainQ31()
	{
		delete this->scaleFactor;
	}

	void GainQ31::Filter(__int32* block, __int32 offset)
	{
		__int32 maxSample = offset + Constant::FilterBlockSizeInInt32s;
		for (__int32 sample = offset; sample < maxSample; ++sample)
		{
			block[sample] = (__int32)((*(this->scaleFactor) * block[sample]) >> this->scaleFactor->FractionalBits());
		}
	}

	void GainQ31::FilterReverse(__int32* block, __int32 offset)
	{
		this->Filter(block, offset);
	}
}
